File: apps/android-companion/app/src/main/java/com/mtog/app/session/SessionReplayGuard.kt
package com.mtog.app.session

class SessionReplayGuard {
    private val highestSequenceBySession = mutableMapOf<String, Long>()

    fun reset() {
        highestSequenceBySession.clear()
    }

    fun accept(message: SessionEnvelope): Boolean {
        val previous = highestSequenceBySession[message.sessionId]
        if (previous != null && message.sequenceNo <= previous) {
            return false
        }

        highestSequenceBySession[message.sessionId] = message.sequenceNo
        return true
    }
}

File: Sources/MtoGMac/SessionReplayGuard.swift
import Foundation

final class SessionReplayGuard {
    private var highestSequenceBySession: [String: UInt64] = [:]

    func reset() {
        highestSequenceBySession.removeAll()
    }

    func accept(_ message: SessionEnvelope) -> Bool {
        let lastSeen = highestSequenceBySession[message.sessionId]
        if let lastSeen, message.sequenceNo <= lastSeen {
            return false
        }

        highestSequenceBySession[message.sessionId] = message.sequenceNo
        return true
    }
}
